peerstore/src/json_peerstore.rs

```diff
@@ -80,6 +81,45 @@ impl<'a> Peerstore for &'a JsonPeerstore {
         JsonPeerstoreAccess(self.store.lock_or_create(hash.into()))
     }
 
+    fn random_peer(self) -> Option<(PeerId, Self::PeerAccess)> {
+        // Since everything can be racy, we use a loop and continue until everything works.
+        loop {
+            let store_len = self.store.len();
+            if store_len == 0 {
+                return None;
+            }
+
+            let num = rand::random::<usize>() % store_len;
+
+            let query = self.store.query(Query {
+                prefix: "".into(),
+                filters: vec![],
+                orders: vec![],
+                skip: num as u64,
+                limit: 1,
+                keys_only: true,
+            });
+
+            // Wait can never block for the JSON datastore.
+            let elem = match query.into_future().wait().ok() {
+                Some((Some((e, _)), _)) => e,
+                _ => continue
+            };
+
+            let id = match PeerId::from_bytes(bs58::decode(&elem).into_vec().ok()?).ok() {
+                Some(id) => id,
+                None => continue
+            };
+
+            let access = match self.store.lock(elem.into()).map(JsonPeerstoreAccess) {
+                Some(acc) => acc,
+                None => continue
+            };
```

Comment:
@twittner the number of items skipped is a random value from `0-store.len()` (unless the store is empty in which case this immediately returns). In reality this loop will probably terminate in one or two cicles. Howerver, I am bit concerned that this might accidentally be used from within a code that already holds a lock on `self.store`, thus leading to an endless loop.

How, instead of an `Option`, this returns a `Result` and loops only ... 10 times max and fails after to prevent deadlocking and endless looping?

Reply:
What I had in mind is that if the decoding of `PeerId` fails the `continue` would retry and if the store contains only one element it would retry the same faulty data over and over again.

Reply:
@twittner I see now. Yeah, I think limiting the loop count is a good idea for that case, too. Although I wonder, why we should have unparsable PeerId-Entries in the store in the first place. Can't we avoid that?